rclcpp/include/rclcpp/executor.hpp

```diff
@@ -381,6 +383,47 @@ class Executor
     return FutureReturnCode::INTERRUPTED;
   }
 
+  /// Spin (blocking) until it times out, or rclcpp is interrupted.
+  /**
+   * \param[in] timeout Timeout parameter, which gets passed to Executor::spin_node_once.
+   * \throws std::runtime_error if spinning is already taking place.
+   */
+  template<typename TimeRepT = int64_t, typename TimeT = std::milli>
+  void spin_until_timeout(
+    std::chrono::duration<TimeRepT, TimeT> timeout)
+  {
+    auto end_time = std::chrono::steady_clock::now();
+
+    if (spinning.exchange(true)) {
+      throw std::runtime_error("spin_until_timeout() called while already spinning");
+    }
+
+    std::chrono::nanoseconds timeout_ns = std::chrono::duration_cast<std::chrono::nanoseconds>(
+      timeout);
+    if (timeout_ns <= std::chrono::nanoseconds::zero()) {
+      // No work to be done, timeout is reached.
+      return;
+    }
+
+    end_time += timeout_ns;
+    std::chrono::nanoseconds timeout_left = timeout_ns;
+
+    RCPPUTILS_SCOPE_EXIT(this->spinning.store(false); );
```

Comment:
this must be kept as close as possible to `if (spinning.exchange(true)) {`.
in the current code, if `timeout_ns <= std::chrono::nanoseconds::zero()` is verified, the node will return without setting `spinning` back to false